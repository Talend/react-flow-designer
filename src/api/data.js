/**
 * This module is private and deal with updating a graph object internal Immutable.Map
 */
import curry from 'lodash/curry';
import isString from 'lodash/isString';
import Immutable from 'immutable';

/**
 * return true if the parameter is an Immutable.Map throw otherwise
 * @param {any} map
 * @return {Bool}
 */
export function isMapElseThrow(map) {
	const test = Immutable.Map.isMap(map);
	if (!test) {
		throw new Error(
			`can't process value as the target is not an Immutable.Map was given a ${map &&
				map.toString()}.`,
		);
	}
	return test;
}

/**
 * return true if the parameter key is a String throw otherwise
 * @param {any} map
 * @return {Bool}
 */
export function isKeyElseThrow(key) {
	const test = isString(key);
	if (!test) {
		throw new Error(`key should be a string was given ${key && key.toString()}`);
	}
	return test;
}

/**
 * given a key and a value, add those to a map
 * @param {String} key
 * @param {any} value
 * @param {Immutable.Map} map
 * @returns {Immutable.Map}
 */
const set = curry((key, value, map) => {
	if (isKeyElseThrow(key) && isMapElseThrow(map)) {
		return map.set(key, value);
	}
	return map;
});

/**
 * given a key and a map return the value associated if exist
 * @param {String} key
 * @param {Immutable.Map} map
 * @returns {any | null}
 */
const get = curry((key, map) => {
	if (isKeyElseThrow(key) && isMapElseThrow(map)) {
		return map.get(key);
	}
	return null;
});

/**
 * Given a key and a map check if this key exist on the map
 * @param {String} key
 * @param {Immutable.Map} map
 * @return {Bool}
 */
const has = curry((key, map) => {
	if (isKeyElseThrow(key) && isMapElseThrow(map)) {
		return map.has(key);
	}
	return false;
});

/**
 * @param {String} key
 * @param {Immutable.Map} map
 * @returns {Immutable.Map}
 */
const deleteKey = curry((key, map) => {
	if (isKeyElseThrow(key) && isMapElseThrow(map)) {
		return map.delete(key);
	}
	return map;
});

export const Data = {
	set,
	get,
	has,
	delete: deleteKey,
};
