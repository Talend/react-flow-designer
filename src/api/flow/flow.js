import flow from 'lodash/flow';
import curry from 'lodash/curry';
import Immutable from 'immutable';

import { throwInDev } from '../throwInDev';
import { Node } from './..';

/**
 * check if node exist in flow
 * @param {FlowState} state
 * @param {string} nodeId
 * @return {bool} true if node exist
 */
export const isNodeExist = curry((state, nodeId) => state.hasIn(['nodes', nodeId]));

const setOut = curry((nodeId, state) => state.setIn(['out', nodeId], new Immutable.Map()));
const deleteOut = curry((nodeId, state) => state.deleteIn(['out', nodeId]));

const setIn = curry((nodeId, state) => state.setIn(['in', nodeId], new Immutable.Map()));
const deleteIn = curry((nodeId, state) => state.deleteIn(['in', nodeId]));

const setChildrens = curry((nodeId, state) =>
	state.setIn(['childrens', nodeId], new Immutable.Map()),
);
const deleteChildrens = curry((nodeId, state) => state.deleteIn(['childrens', nodeId]));

const setParents = curry((nodeId, state) => state.setIn(['parents', nodeId], new Immutable.Map()));
const deleteParents = curry((nodeId, state) => state.deleteIn(['parents', nodeId]));

/**
 * add a node to the flow
 * @param {FlowState} state
 * @param {NodeRecord} node
 * @return {FlowState}
 */
export const addNode = curry((state, node) => {
	if (Node.isNodeElseThrow(node) && !isNodeExist(state, Node.getId(node))) {
		const nodeId = Node.getId(node);
		return flow([setOut(nodeId), setIn(nodeId), setChildrens(nodeId), setParents(nodeId)])(
			state.setIn(['nodes', Node.getId(node)], node),
		);
	}
	throwInDev(`Node with id = ${Node.getId(node)}, already exist, can't create node.`);
	return state;
});

/**
 * if exist remove a node from the flow
 * @param {FlowState} state
 * @param {string} nodeId
 * @return {FlowState}
 */
export const deleteNode = curry((state, nodeId) => {
	if (isNodeExist(nodeId)) {
		return flow([
			deleteOut(nodeId),
			deleteIn(nodeId),
			deleteChildrens(nodeId),
			deleteParents(nodeId),
		])(state.deleteIn(['nodes', nodeId]));
	}
	return state;
});

/**
 * update a node
 * @param {FlowState} state
 * @param {string} nodeId
 * @param {NodeRecord} node
 * @return {FlowState}
 */
export const updateNode = curry((state, nodeId, node) => {
	if (Node.isNodeElseThrow(node) && (isNodeExist(state, Node.getId(node)), isNodeExist(nodeId))) {
		if (nodeId === Node.getId(node)) {
			return addNode(deleteNode(state, Node.getId(node)), node);
		}
		return state.setIn(['nodes', nodeId], node);
	}
	return state;
});
