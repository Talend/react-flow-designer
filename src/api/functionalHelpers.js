import curry from 'lodash/curry';

/**
 * given a value an defautl value, act as identity function except if value is undefined
 * returning default value instead.
 * @param {a} defaultValue
 * @param {b} value
 * @return {a | b}
 */
export const withDefault = curry((defaultValue, value) => {
	if (value === undefined) {
		return defaultValue;
	}
	return value;
});

/**
 * if value is undefined, does not apply the function to this value
 * @param { a => b }
 * @param {a}
 * @return {a | b}
 */
export const andThen = curry((func, value) => {
	if (value === undefined) {
		return value;
	}
	return func(value);
});
