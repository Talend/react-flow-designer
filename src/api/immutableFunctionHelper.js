import curry from 'lodash/curry';
import Immutable from 'immutable';

import { throwTypeError } from './throwInDev';

/**
 * this function is here to help using immutable in functionnal programming friendly context
 * @example <caption>Handling Uncertainty in safe way</caption>
 * flow([mapFunction(returnMap), mapFunction(getImmutableMapValue)])
 * @param {any} key
 * @param {Immutable.Map} map
 * @return {any} the value stored in the map or null if map argument is not an Immutable.Map
 */
export const getImmutableMapValue = curry((key, map) => {
	const test = Immutable.Map.isMap(map);
	if (test) {
		return map.get(key, null);
	}
	throwTypeError('Immutable.Map', map, 'map');
	return null;
});

export const setImmutableMapValue = curry((key, value, map) => {
	const test = Immutable.Map.isMap(map);
	if (test) {
		return map.set(key, value);
	}
	throwTypeError('Immutable.Map', map, 'map');
	return map;
});

export const deleteImmutableMap = curry((key, map) => {
	const test = Immutable.Map.isMap(map);
	if (test) {
		return map.delete(key);
	}
	throwTypeError('Immutable.Map', map, 'map');
	return map;
});
