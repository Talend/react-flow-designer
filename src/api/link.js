/**
 * This module is public and deal with Graph's object Links
 */

import curry from 'lodash/curry';
import flow from 'lodash/flow';
import indexOf from 'lodash/indexOf';
import isString from 'lodash/isString';
import upperFirst from 'lodash/upperFirst';

import { throwInDev, throwTypeError } from './throwInDev';
import { LinkRecord } from '../constants/flowdesigner.model';
import * as Data from './data';

const linkTypeSelector = ['graphicalAttributes', 'linkType'];

/** in future properties should be removed from the react-flow-designer lib */
const FORBIDEN_GRAPHICAL_ATTRIBUTES = ['properties', 'linkType'];

/**
 * Test if the first parameter is a LinkRecord instance
 * @param {LinkRecord} link
 * @returns {bool}
 * @throws
 */
function isLink(link) {
	if (link && link instanceof LinkRecord) {
		return true;
	}
	return false;
}

/**
 * Test if the first parameter is a LinkRecord, throw if not
 * @param {*} link
 * @returns {bool}
 * @throws
 */
export function isLinkElseThrow(link) {
	const test = isLink(link);
	if (!test) {
		throwTypeError('Linkrecord', link, 'link', 'Link');
	}
	return test;
}

/**

 * @param {LinkRecord} link
 * @return {string}
 */
function getId(link) {
	if (isLinkElseThrow(link)) {
		return link.get('id');
	}
	return false;
}

/**
 * @param {string} id
 * @param {LinkRecord} link
 * @returns {LinkRecord}
 */
const setId = curry((id, link) => {
	if (isString(id) && isLinkElseThrow(link)) {
		return link.set('id', id);
	}
	throwInDev(`id should be a string was given ${id && id.toString()}`);
	return link;
});

/**
 * @param {LinkRecord} link
 * @returns {string}
 */
function getSourceId(link) {
	if (isLinkElseThrow(link)) {
		return link.get('sourceId');
	}
	return false;
}

/**
 * @param {string} sourceId
 * @param {LinkRecord} link
 * @returns {LinkRecord}
 */
const setSourceId = curry((sourceId, link) => {
	if (isString(sourceId) && isLinkElseThrow(link)) {
		return link.set('sourceId', sourceId);
	}
	throwInDev(`id should be a string was given ${sourceId && sourceId.toString()}`);
	return link;
});

/**
 * @param {LinkRecord} link
 * @returns {string}
 */
function getTargetId(link) {
	if (isLinkElseThrow(link)) {
		return link.get('targetId');
	}
	return false;
}

/**
 * @param {string} targetId
 * @param {LinkRecord} link
 * @returns {LinkRecord}
 */
const setTargetId = curry((targetId, link) => {
	if (isString(targetId) && isLinkElseThrow(link)) {
		return link.set('targetId', targetId);
	}
	throwInDev(`id should be a string was given ${targetId && targetId.toString()}`);
	return link;
});

/**
 * @param {LinkRecord} link
 * @returns {LinkRecord}
 */
function getComponentType(link) {
	if (isLinkElseThrow(link, true)) {
		return link.getIn(linkTypeSelector);
	}
	return false;
}

/**
 * @param {string} linkType
 * @param {LinkRecord} link
 * @returns {LinkRecord}
 */
const setComponentType = curry((linkType, link) => {
	if (isString(linkType) && isLinkElseThrow(link, true)) {
		return link.setIn(linkTypeSelector, linkType);
	}
	throwInDev(`linkType should be a string was given ${linkType && linkType.toString()}`);
	return link;
});

/**
 * @param {String} key
 * @param {any} value
 * @param {nodeRecord} node
 * @returns {nodeRecord}
 */
const setData = curry((key, value, link) => {
	if (isLinkElseThrow(link)) {
		return link.set('data', Data.set(key, value, link.get('data')));
	}
	return link;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {any | null}
 */
const getData = curry((key, link) => {
	if (isLinkElseThrow(link)) {
		return Data.get(key, link.get('data'));
	}
	return null;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {Bool}
 */
const hasData = curry((key, link) => {
	if (isLinkElseThrow(link)) {
		return Data.has(key, link.get('data'));
	}
	return false;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const deleteData = curry((key, link) => {
	if (isLinkElseThrow(link)) {
		return link.set('data', Data.deleteKey(key, link.get('data')));
	}
	return link;
});

/**
 * given a key check if that key is white listed
 * @param {String} key
 * @returns {Bool}
 */
function isWhiteListAttribute(key) {
	if (indexOf(FORBIDEN_GRAPHICAL_ATTRIBUTES, key) === -1) {
		return true;
	}
	throwInDev(
		`${key} is a protected value of the Link, please use get${upperFirst(key)} set${upperFirst(
			key,
		)} from this module to make change on those values`,
	);
	return false;
}

/**
 * @param {String} key
 * @param {any} value
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const setGraphicalAttribute = curry((key, value, link) => {
	if (isLinkElseThrow(link) && isWhiteListAttribute(key)) {
		return link.set(
			'graphicalAttributes',
			Data.set(key, value, link.get('graphicalAttributes')),
		);
	}
	return link;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {any | null}
 */
const getGraphicalAttribute = curry((key, link) => {
	if (isLinkElseThrow(link) && isWhiteListAttribute(key)) {
		return Data.get(key, link.get('graphicalAttributes'));
	}
	return null;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {Bool}
 */
const hasGraphicalAttribute = curry((key, link) => {
	if (isLinkElseThrow(link) && isWhiteListAttribute(key)) {
		return Data.has(key, link.get('graphicalAttributes'));
	}
	return false;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const deleteGraphicalAttribute = curry((key, link) => {
	if (isLinkElseThrow(link) && isWhiteListAttribute(key)) {
		return link.set('graphicalAttributes', Data.deleteKey(key, link.get('graphicalAttributes')));
	}
	return link;
});

/**
 * minimal link creation factory, additionnals information can be set trought
 * the above set* functions
 * @param {string} id
 * @param {string} sourceId
 * @param {string} targetId
 * @param {string} componenttype
 * @return {LinkRecord}
 */
const create = curry((id, sourceId, targetId, componentType) =>
	flow([
		setId(id),
		setSourceId(sourceId),
		setTargetId(targetId),
		setComponentType(componentType),
	])(new LinkRecord()),
);

export const Link = {
	create,
	isLink,
	getId,
	getSourceId,
	setSourceId,
	getTargetId,
	setTargetId,
	getComponentType,
	setComponentType,
	setData,
	getData,
	hasData,
	deleteData,
	setGraphicalAttribute,
	getGraphicalAttribute,
	hasGraphicalAttribute,
	deleteGraphicalAttribute,
};
