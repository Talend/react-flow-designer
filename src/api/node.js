/**
 * This module is public and deal with Graph's object Nodes
 */
import curry from 'lodash/curry';
import flow from 'lodash/flow';
import indexOf from 'lodash/indexOf';
import isString from 'lodash/isString';
import upperFirst from 'lodash/upperFirst';

import throwInDev from './throwInDev';
import { NodeRecord } from '../constants/flowdesigner.model';
import { isPositionElseThrow } from './position';
import { isSizeElseThrow } from './size';
import { Data } from './data';

const positionSelector = ['graphicalAttributes', 'position'];
const sizeSelector = ['graphicalAttributes', 'nodeSize'];
const componentTypeSelector = ['graphicalAttributes', 'nodeType'];

const FORBIDEN_GRAPHICAL_ATTRIBUTES = ['position', 'nodeSize', 'nodeType'];

/**
 * Test if the first parameter is a NodeRecord instance
 * @param {NodeRecord} node
 * @param {bool} doThrow - throw if not a node
 * @returns {bool}
 * @throws
 */
function isNode(node) {
	if (node && node instanceof NodeRecord) {
		return true;
	}
	return false;
}

/**
 * Test if the first parameter is a NodeRecord, throw if not
 * @param {*} node
 * @returns {bool}
 * @throws
 */
export function isNodeElseThrow(node) {
	const test = isNode(node);
	if (!test) {
		throwInDev(
			`Should be a NodeRecord was given ${node &&
				node.toString()}, you should use Node module functions to create and transform Nodes`,
		);
	}
	return test;
}

/**
 * @param {NodeRecord} node
 * @returns {string}
 */
function getId(node) {
	if (isNodeElseThrow(node)) {
		return node.get('id');
	}
	return null;
}

/**
 * @param {string} id
 * @param {NodeRecord}
 * @returns {NodeRecord}
 */
const setId = curry((id, node) => {
	if (isString(id) && isNodeElseThrow(node)) {
		return node.set('id', id);
	}
	throwInDev(`nodeId should be a string was given ${id && id.toString()}}`);
	return node;
});

/**
 * @param {NodeRecord} node
 * @returns {PositionRecord}
 */
function getPosition(node) {
	if (isNodeElseThrow(node)) {
		return node.getIn(positionSelector);
	}
	return null;
}

/**
 * @param {PositionRecord} position
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const setPosition = curry((position, node) => {
	if (isPositionElseThrow(position) && isNodeElseThrow(node)) {
		return node.setIn(positionSelector, position);
	}
	return node;
});

/**
 * @param {NodeRecord} node
 * @returns {PositionRecord}
 */
function getSize(node) {
	if (isNodeElseThrow(node)) {
		return node.getIn(sizeSelector);
	}
	return null;
}

/**
 * @param {SizeRecord} size
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const setSize = curry((size, node) => {
	if (isSizeElseThrow(size) && isNodeElseThrow(node)) {
		return node.setIn(sizeSelector, size);
	}
	return node;
});

function getComponentType(node) {
	if (isNodeElseThrow(node)) {
		return node.getIn(componentTypeSelector);
	}
	return null;
}

const setComponentType = curry((nodeType, node) => {
	if (isString(nodeType) && isNodeElseThrow(node)) {
		return node.setIn(componentTypeSelector, nodeType);
	}
	throwInDev(`nodeType should be a string was given ${nodeType && nodeType.toString()}`);
	return node;
});

/**
 * @param {String} key
 * @param {any} value
 * @param {nodeRecord} node
 * @returns {nodeRecord}
 */
const setData = curry((key, value, node) => {
	if (isNodeElseThrow(node)) {
		return node.set('data', Data.set(key, value, node.get('data')));
	}
	return node;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {any | null}
 */
const getData = curry((key, node) => {
	if (isNodeElseThrow(node)) {
		return Data.get(key, node.get('data'));
	}
	return null;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {Bool}
 */
const hasData = curry((key, node) => {
	if (isNodeElseThrow(node)) {
		return Data.has(key, node.get('data'));
	}
	return false;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const deleteData = curry((key, node) => {
	if (isNodeElseThrow(node)) {
		return node.set('data', Data.delete(key, node.get('data')));
	}
	return node;
});

/**
 * given a key check if that key is white listed
 * @param {String} key
 * @returns {Bool}
 */
function isWhiteListAttribute(key) {
	if (indexOf(FORBIDEN_GRAPHICAL_ATTRIBUTES, key)) {
		return true;
	}
	throwInDev(
		`${key} is a protected value of the Node, please use get${upperFirst(key)} set${upperFirst(
			key,
		)} from this module to make change on those values`,
	);
	return false;
}

/**
 * @param {String} key
 * @param {any} value
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const setGraphicalAttribute = curry((key, value, node) => {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return node.set(
			'graphicalAttributes',
			Data.set(key, value, node.get('graphicalAttributes')),
		);
	}
	return node;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {any | null}
 */
const getGraphicalAttribute = curry((key, node) => {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return Data.get(key, node.get('graphicalAttributes'));
	}
	return null;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {Bool}
 */
const hasGraphicalAttribute = curry((key, node) => {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return Data.has(key, node.get('graphicalAttributes'));
	}
	return false;
});

/**
 * @param {String} key
 * @param {NodeRecord} node
 * @returns {NodeRecord}
 */
const deleteGraphicalAttribute = curry((key, node) => {
	if (isNodeElseThrow(node) && isWhiteListAttribute(key)) {
		return node.set('graphicalAttributes', Data.delete(key, node.get('graphicalAttributes')));
	}
	return node;
});

/**
 * Create a new Node
 * @param {String} id
 * @param {PositionRecord} position
 * @param {SizeRecord} size
 * @param {String} componentType
 * @returns {NodeRecord}
 */
const create = curry((id, position, size, componentType) =>
	flow([setId(id), setPosition(position), setSize(size), setComponentType(componentType)])(
		new NodeRecord(),
	),
);

export const Node = {
	create,
	isNode,
	getId,
	getPosition,
	setPosition,
	getSize,
	setSize,
	getComponentType,
	setComponentType,
	setData,
	getData,
	hasData,
	deleteData,
	setGraphicalAttribute,
	getGraphicalAttribute,
	hasGraphicalAttribute,
	deleteGraphicalAttribute,
};
