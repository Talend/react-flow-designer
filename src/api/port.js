import curry from 'lodash/curry';
import flow from 'lodash/flow';
import Immutable from 'immutable';
import { PortRecord, PositionRecord } from '../constants/flowdesigner.model';

const positionSelector = ['graphicalAttributes', 'position'];
const componentTypeSelector = ['graphicalAttributes', 'portType'];
const portTopologySelector = ['graphicalAttributes', 'properties', 'type'];
const indexSelector = ['graphicalAttributes', 'properties', 'index'];

export function isPositionRecord(position, doThrow) {
	if (position && position instanceof PositionRecord) {
		return true;
	}
	if (doThrow) {
		throw new Error(`Should be a PositionRecord was given ${position.toString()}`);
	}
	return false;
}

/**
 * Test if the first parameter is a PortRecord
 * @param {*} port
 * @param {bool} doThrow - throw if not a port
 * @return {bool}
 * @throws
 */
export function isPortRecord(port, doThrow = false) {
	if (port && port instanceof PortRecord) {
		return true;
	}
	if (doThrow) {
		throw new Error(`Should be a PortRecord was given ${port.toString()}`);
	}
	return false;
}

/**
 * Test if the first parameter is a PortRecord, throw if not
 * @param {*} port
 * @return {bool}
 * @throws
 */
export function isPortRecordElseThrow(port) {
	return isPortRecord(port, true);
}

export function isTypology(typology, doThrow) {
	if (typology === 'SOURCE' || typology === 'SINK') {
		return true;
	}
	if (doThrow) {
		throw new Error(`Should be a typology 'SOURCE' or 'SINK' was given ${typology.toString()}`);
	}
	return false;
}

export function getId(port) {
	const whatver = isPortRecordElseThrow(port);
	if (whatver) {
		return port.get('id');
	}
	return false;
}

const setId = curry((id, port) => {
	if (isPortRecord(port) && typeof id === 'string') {
		return port.set('id', id);
	}
	throw new Error(`id should be a string was given ${id.toString()}`);
});

export function getNodeId(port) {
	if (isPortRecord(port, true)) {
		return port.get('nodeId');
	}
	return false;
}

export const setNodeId = curry((nodeId, port) => {
	if (isPortRecord(port, true) && typeof nodeId === 'string') {
		return port.set('nodeId', nodeId);
	}
	throw new Error(`nodeId should be a string was given ${nodeId.toString()}`);
});

export function getPosition(port) {
	if (isPortRecord(port, true)) {
		return port.getIn(positionSelector);
	}
	return false;
}

export const setPosition = curry((position, port) => {
	if (isPortRecord(port, true) && isPositionRecord(position, true)) {
		return port.setIn(positionSelector, position);
	}
	return false;
});

export function getComponentType(port) {
	if (isPortRecord(port, true)) {
		return port.getIn(componentTypeSelector);
	}
	return false;
}

export const setComponentType = curry((componentType, port) => {
	if (isPortRecord(port, true) && typeof componentType === 'string') {
		return port.setIn(componentTypeSelector, componentType);
	}
	throw new Error(`componentType should be a string was given ${componentType.toString()}`);
});

export function getTypology(port) {
	if (isPortRecord(port, true)) {
		return port.getIn(portTopologySelector);
	}
	return false;
}
export const setTypology = curry((flowType, port) => {
	if (isPortRecord(port, true && isTypology(flowType))) {
		return port.setIn(portTopologySelector, flowType);
	}
	return false;
});
export function getIndex(port) {
	if (isPortRecord(port, true)) {
		return port.getIn(indexSelector);
	}
	return false;
}
export const setIndex = curry((index, port) => {
	if (isPortRecord(port, true) && typeof index === 'number') {
		return port.setIn(indexSelector, index);
	}
	throw new Error(`index should be a number was given ${index.toString()}`);
});

export function getData(port) {
	if (isPortRecord(port)) {
		return port.get('data');
	}
	return false;
}

export const setData = curry((map, port) => {
	if (isPortRecord(port) && Immutable.Map.isMap(map)) {
		return port.set('data', map);
	}
	throw new Error(`data should be a Immutable.Map go ${map.toString()}`);
});

export const createPortRecord = curry((id, nodeId, index, typology) => {
	const create = flow([setId(id), setNodeId(nodeId), setIndex(index), setTypology(typology)]);
	return create(new PortRecord());
});
