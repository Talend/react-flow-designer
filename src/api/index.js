import { Port as port } from './port';
import { Link as link } from './link';
import { Node as node } from './node';
import { Position as position } from './position';
import { Size as size } from './size';
import * as data from './data';

export const Port = port;
export const Link = link;
export const Node = node;
export const Position = position;
export const Size = size;
export const Data = data;
