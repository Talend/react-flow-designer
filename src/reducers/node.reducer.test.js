import { Map } from 'immutable';

import { defaultState } from './flow.reducer';
import nodeReducer from './node.reducer';
import {
	NodeRecord,
	PositionRecord,
	NodeGraphicalAttributes,
} from '../constants/flowdesigner.model';
import {
	FLOWDESIGNER_NODE_SET_TYPE,
	FLOWDESIGNER_NODE_MOVE,
	FLOWDESIGNER_NODE_MOVE_END,
} from '../constants/flowdesigner.constants';
import { Node, Position, Size } from '../api';

describe('Check node reducer', () => {
	const initialState = defaultState
		.setIn(
			['nodes', 'id1'],
			new NodeRecord({
				id: 'id1',
				type: 'type1',
				data: new Map({ type: 'test' }),
				graphicalAttributes: new NodeGraphicalAttributes({
					type: 'type1',
					selected: true,
					position: new PositionRecord({ x: 10, y: 10 }),
				}),
			}),
		)
		.setIn(
			['nodes', 'id2'],
			new NodeRecord({
				id: 'id2',
				type: 'type2',
				data: new Map({ type: 'test' }),
				graphicalAttributes: new NodeGraphicalAttributes({
					type: 'type2',
					selected: false,
					position: new PositionRecord({ x: 10, y: 10 }),
				}),
			}),
		);

	it('FLOWDESIGNER_NODE_ADD properly add a new node to the node collection', () => {
		expect(
			nodeReducer(defaultState, {
				type: 'FLOWDESIGNER_NODE_ADD',
				nodeId: 'id',
				graphicalAttributes: {
					position: { x: 10, y: 10 },
				},
			}),
		).toMatchSnapshot();
	});

	it('FLOWDESIGNER_NODE_ADD add a new node to the node collection with the right type', () => {
		expect(
			nodeReducer(defaultState, {
				type: 'FLOWDESIGNER_NODE_ADD',
				nodeId: 'id',
				graphicalAttributes: {
					name: 'test',
					position: { x: 10, y: 10 },
					nodeSize: { height: 50, width: 50 },
					type: 'MY_NODE_TYPE',
				},
			}),
		).toMatchSnapshot();
		expect(
			nodeReducer(defaultState, {
				type: 'FLOWDESIGNER_NODE_ADD',
				node: Node.create(
					'id',
					Position.create(10, 10),
					Size.create(50, 50),
					'MY_NODE_TYPE',
				),
			}),
		).toMatchSnapshot();
	});

	it('FLOWDESIGNER_NODE_MOVE update node position', () => {
		expect(
			nodeReducer(initialState, {
				type: FLOWDESIGNER_NODE_MOVE,
				nodeId: 'id2',
				nodePosition: { x: 50, y: 50 },
			}),
		).toMatchSnapshot();
	});

	it('empty the startPostion when receiving a FLOW_DESIGNER_MOVE_END command', () => {
		expect(
			nodeReducer(initialState, {
				type: FLOWDESIGNER_NODE_MOVE_END,
				nodeId: 'id1',
				nodePosition: { x: 50, y: 50 },
			}).getIn(['nodes', 'id1', 'graphicalAttributes', 'properties', 'startPosition']),
		).toEqual(undefined);
	});

	it('FLOWDESIGNER_NODE_SET_SIZE update node size property', () => {
		expect(
			nodeReducer(initialState, {
				type: 'FLOWDESIGNER_NODE_SET_SIZE',
				nodeId: 'id1',
				nodeSize: { height: 200, width: 200 },
			}),
		).toMatchSnapshot();
	});

	it('FLOWDESIGNER_NODE_SET_TYPE update node type', () => {
		const nodeId = 'id1';
		const nodeType = 'nodetype';

		expect(
			nodeReducer(initialState, {
				type: FLOWDESIGNER_NODE_SET_TYPE,
				nodeId,
				nodeType,
			}),
		).toMatchSnapshot();
	});

	it('FLOWDESIGNER_NODE_SET_GRAPHICAL_ATTRIBUTES should add { selected: false } attribute to node graphicalAttributes map', () => {
		expect(
			nodeReducer(initialState, {
				type: 'FLOWDESIGNER_NODE_SET_GRAPHICAL_ATTRIBUTES',
				nodeId: 'id1',
				graphicalAttributes: { selected: false },
			}),
		).toMatchSnapshot();
	});

	it('FLOWDESIGNER_NODE_REMOVE_GRAPHICAL_ATTRIBUTES should remove {selected} attribute to node graphicalAttributes map', () => {
		expect(
			nodeReducer(initialState, {
				type: 'FLOWDESIGNER_NODE_REMOVE_GRAPHICAL_ATTRIBUTES',
				nodeId: 'id1',
				graphicalAttributesKey: 'selected',
			}),
		).toMatchSnapshot();
	});

	it("FLOWDESIGNER_NODE_SET_DATA should add { type: 'string' } attribute to node data map", () => {
		expect(
			nodeReducer(initialState, {
				type: 'FLOWDESIGNER_NODE_SET_DATA',
				nodeId: 'id1',
				data: { type: 'string' },
			}),
		).toMatchSnapshot();
	});

	it('FLOWDESIGNER_NODE_REMOVE_DATA should remove {type} attribute to node data map', () => {
		expect(
			nodeReducer(initialState, {
				type: 'FLOWDESIGNER_NODE_REMOVE_DATA',
				nodeId: 'id1',
				data: 'type',
			}),
		).toMatchSnapshot();
	});

	it('FLOWDESIGNER_NODE_REMOVE should remove node from node collection', () => {
		expect(
			nodeReducer(initialState, {
				type: 'FLOWDESIGNER_NODE_REMOVE',
				nodeId: 'id1',
			}),
		).toMatchSnapshot();
	});
});

describe('FLOWDESIGNER_NODE_APPLY_MOVEMENT', () => {
	const initialState = defaultState
		.setIn(
			['nodes', 'id1'],
			new NodeRecord({
				id: 'id1',
				nodeType: 'type1',
				graphicalAttributes: new NodeGraphicalAttributes({
					position: new PositionRecord({ x: 10, y: 10 }),
				}),
			}),
		)
		.setIn(
			['nodes', 'id2'],
			new NodeRecord({
				id: 'id2',
				nodeType: 'type2',
				graphicalAttributes: new NodeGraphicalAttributes({
					position: new PositionRecord({ x: 10, y: 10 }),
				}),
			}),
		)
		.setIn(
			['nodes', 'id3'],
			new NodeRecord({
				id: 'id3',
				nodeType: 'type2',
				graphicalAttributes: new NodeGraphicalAttributes({
					position: new PositionRecord({ x: 10, y: 10 }),
				}),
			}),
		);
	it('should apply the same relative movement to each node listed', () => {
		expect(
			nodeReducer(initialState, {
				type: 'FLOWDESIGNER_NODE_APPLY_MOVEMENT',
				nodesId: ['id1', 'id2'],
				movement: { x: 10, y: 5 },
			}),
		).toMatchSnapshot();
	});
});
