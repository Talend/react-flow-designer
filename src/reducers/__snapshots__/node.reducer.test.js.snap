exports[`Check node reducer FLOWDESIGNER_NODE_ADD add a new node to the node collection with the right type 1`] = `
Object {
  "in": Object {
    "id": Object {},
  },
  "nodes": Object {
    "id": Object {
      "attributes": Object {
        "name": "test",
      },
      "id": "id",
      "nodeSize": Object {
        "height": undefined,
        "width": undefined,
      },
      "nodeType": "MY_NODE_TYPE",
      "position": Object {
        "x": 10,
        "y": 10,
      },
    },
  },
  "out": Object {
    "id": Object {},
  },
  "preds": Object {
    "id": Object {},
  },
  "sucs": Object {
    "id": Object {},
  },
}
`;

exports[`Check node reducer FLOWDESIGNER_NODE_ADD properly add a new node to the node collection 1`] = `
Object {
  "in": Object {
    "id": Object {},
  },
  "nodes": Object {
    "id": Object {
      "attributes": Object {},
      "id": "id",
      "nodeSize": Object {
        "height": undefined,
        "width": undefined,
      },
      "nodeType": undefined,
      "position": Object {
        "x": 10,
        "y": 10,
      },
    },
  },
  "out": Object {
    "id": Object {},
  },
  "preds": Object {
    "id": Object {},
  },
  "sucs": Object {
    "id": Object {},
  },
}
`;

exports[`Check node reducer FLOWDESIGNER_NODE_REMOVE should remove node from node collection 1`] = `
Object {
  "nodes": Object {
    "id2": Object {
      "attributes": Object {
        "selected": false,
      },
      "id": "id2",
      "nodeSize": undefined,
      "nodeType": "type2",
      "position": Object {
        "x": 10,
        "y": 10,
      },
    },
  },
}
`;
